### PR TITLE
Metadata uploader validate file

### PR DESCRIPTION
**Pull request by ksierks**
## Description of changes
Display a validation error when the file is invalid (ie. duplicate or empty header names).
image

## Related issue
N/A

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* ~[ ] CHANGELOG.md (and UPGRADING.md if necessary) updated with information for new change.~
* [x] Tests added (or description of how to test) for any new features.
* ~[ ] User documentation updated for UI or technical changes.~

### CODE REVIEW

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx [hunk 1]
```diff
@@ -25,61 +25,102 @@ export function SampleMetadataImportSelectFile(): JSX.Element {
...
       }
-      return false;
+      return true;
     },
   };
 
   return (
     <SampleMetadataImportWizard current={0} status={status}>
       <Spin spinning={loading}>
+        {!valid && (
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx:116]: Add spacing between the `ErrorAlert` and the `DragUpload`. Probably use `marginBottom: SPACE_XS` on the `ErrorAlert`

File: src/test/java/ca/corefacility/bioinformatics/irida/ria/integration/projects/ProjectSampleMetadataImportPageIT.java [hunk 2]
```diff
@@ -89,4 +90,11 @@ public void testFailedUpload() {
 		page.goToReviewPage();
 		assertTrue(page.isAlertDisplayed(), "Validation message did not display");
 	}
+
+	@Test
+	public void testDuplicateHeaders() {
+		ProjectSampleMetadataImportPage page = ProjectSampleMetadataImportPage.goToPage(driver());
+		page.uploadMetadataFile(DUPLICATE_HEADERS_FILE_PATH);
+		assertTrue(page.isAlertDisplayed(), "Validation message did not display");
+	}
```
Comment (deepsidhu85) [inline src/test/java/ca/corefacility/bioinformatics/irida/ria/integration/projects/ProjectSampleMetadataImportPageIT.java:99]: Add a test for displaying the validation error alert when there are duplicate headers as well.

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx [hunk 3]
```diff
@@ -25,61 +25,102 @@ export function SampleMetadataImportSelectFile(): JSX.Element {
...
+        const data = await readFileContents(file);
+        const workbook: XLSX.WorkBook = XLSX.read(data, {
+          type: "binary",
+          raw: true,
+        });
+        const { SheetNames } = workbook;
+        const [firstSheet] = SheetNames;
+        // Not loving how I'm parsing the file twice to get the raw headers
+        // See https://docs.sheetjs.com/docs/api/utilities
+        const headers: string[] = XLSX.utils.sheet_to_json(
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx:71]: As discussed, update this to only convert the sheet to json once

File: src/main/resources/i18n/messages.properties [hunk 4]
```diff
@@ -1993,6 +1993,8 @@ SampleMetadataImportSelectFile.warning=Metadata uploaded will overwrite duplicat
 SampleMetadataImportSelectFile.dropzone=Click or drop Excel/CSV file containing metadata for samples in this project.
 SampleMetadataImportSelectFile.success={0} was read successfully.
 SampleMetadataImportSelectFile.error={0} failed to upload.
+SampleMetadataImportSelectFile.alert.valid.title=Validation Error
+SampleMetadataImportSelectFile.alert.valid.description=There are duplicate header names in file. Please correct the header names in the file and re-upload.
```
Comment (deepsidhu85) [inline src/main/resources/i18n/messages.properties:1997]: Change this text to: `Duplicate header names were found in the uploaded file. Please correct the header names in the file and then re-upload.`

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
Looks good just a few things below. Also, I'm thinking maybe we should have server side checking for duplicate header names as well. Thoughts @ksierks @ericenns

File: src/test/java/ca/corefacility/bioinformatics/irida/ria/integration/projects/ProjectSampleMetadataImportPageIT.java (hunk 2 above)
Comment (ksierks) [reply src/test/java/ca/corefacility/bioinformatics/irida/ria/integration/projects/ProjectSampleMetadataImportPageIT.java:99]: I'm confused, this is a test for displaying the alert when there are duplicate headers.

File: src/main/resources/i18n/messages.properties (hunk 4 above)
Comment (ksierks) [reply src/main/resources/i18n/messages.properties]: Please see 85d4aa0732ecc77f2ab86d91404aec941445b095.

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx (hunk 1 above)
Comment (ksierks) [reply src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx:117]: Please see 3dbcfcc176217a7368307fb3ea8bd4d351ec2105.

File: src/test/java/ca/corefacility/bioinformatics/irida/ria/integration/projects/ProjectSampleMetadataImportPageIT.java (hunk 2 above)
Comment (deepsidhu85) [reply src/test/java/ca/corefacility/bioinformatics/irida/ria/integration/projects/ProjectSampleMetadataImportPageIT.java:99]: My bad. The message just needs to be updated. Currently it says "Validation messages did not display"

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx (hunk 3 above)
Comment (ksierks) [reply src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx:72]: Tried doing what we discussed, but it didn't work because when
`const rows: MetadataItem[] = XLSX.utils.sheet_to_json(workbook.Sheets[firstSheet], {rawNumbers: false,});`
is called, the excel document is stored in a key-value hash. Key-value hashes cannot have duplicate header names, their keys must be unique. This function automatically alters the headers names making them unique and then I cannot find unique names later.

File: src/test/java/ca/corefacility/bioinformatics/irida/ria/integration/projects/ProjectSampleMetadataImportPageIT.java (hunk 2 above)
Comment (deepsidhu85) [reply src/test/java/ca/corefacility/bioinformatics/irida/ria/integration/projects/ProjectSampleMetadataImportPageIT.java:99]: Disregard this comment

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx (hunk 3 above)
Comment (deepsidhu85) [reply src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx]: Thanks for updating this! :)

**Review by deepsidhu85 — APPROVED** [review]

**Comment by ksierks** [comment]
> Looks good just a few things below. Also, I'm thinking maybe we should have server side checking for duplicate header names as well. Thoughts @ksierks @ericenns

Server cannot create duplicate headers. I added a db constraint that in the base branch already. See [MetadataTemplateField.java](https://github.com/phac-nml/irida/compare/development...metadata-uploader-base#diff-ee3d99e467281cd3ce7a0a95c85387e7e022fa305e5640e2db0e1f48bd7f8ed3).

**Comment by deepsidhu85** [comment]
> > Looks good just a few things below. Also, I'm thinking maybe we should have server side checking for duplicate header names as well. Thoughts @ksierks @ericenns
> 
> Server cannot create duplicate headers. I added a db constraint that in the base branch already. See [MetadataTemplateField.java](https://github.com/phac-nml/irida/compare/development...metadata-uploader-base#diff-ee3d99e467281cd3ce7a0a95c85387e7e022fa305e5640e2db0e1f48bd7f8ed3).

Awesome thanks! Good to go from my end then!

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx [hunk 5]
```diff
@@ -26,7 +26,10 @@ export function SampleMetadataImportSelectFile(): JSX.Element {
   const dispatch: ImportDispatch = useImportDispatch();
   const [status, setStatus] = React.useState<StepsProps["status"]>("process");
   const [loading, setLoading] = React.useState<boolean>(false);
-  const [valid, setValid] = React.useState<boolean>(true);
+  const [showEmptyHeaderError, setShowEmptyHeaderError] =
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx:29]: Instead of using these constants lets replace it with a single one such as `const [uploadValidationErrors, setUploadValidationErrors] = React.useState<ReactElement[]>([]);` Then below we would update line 101 and 104 to push the list item with the relevant text into an array  (such as const validationErrors = []). After line 106 we setUploadValidationErrors(validationErrors).

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx [hunk 6]
```diff
@@ -120,20 +130,42 @@ export function SampleMetadataImportSelectFile(): JSX.Element {
...
       }
-      return true;
     },
   };
 
   return (
     <SampleMetadataImportWizard current={0} status={status}>
       <Spin spinning={loading}>
-        {!valid && (
+        {(showEmptyHeaderError || showDuplicateHeaderError) && (
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx:139]: With the changes made above we can update this section to be:

{uploadValidationErrors.length && (
          <ErrorAlert
            style={{ marginBottom: SPACE_XS }}
            message={i18n("SampleMetadataImportSelectFile.alert.valid.title")}
            description={
              <>
                {i18n(
                  "SampleMetadataImportSelectFile.alert.valid.description.preface"
                )}
                <ul style={{ marginBottom: 0 }}>{uploadValidationErrors}</ul>
                {i18n(
                  "SampleMetadataImportSelectFile.alert.valid.description.postface"
                )}
              </>
            }
          />
        )}

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx [hunk 7]
```diff
@@ -120,20 +130,42 @@ export function SampleMetadataImportSelectFile(): JSX.Element {
...
           <ErrorAlert
             style={{ marginBottom: SPACE_XS }}
             message={i18n("SampleMetadataImportSelectFile.alert.valid.title")}
-            description={i18n(
-              "SampleMetadataImportSelectFile.alert.valid.description"
-            )}
+            description={
+              <>
+                {i18n(
+                  "SampleMetadataImportSelectFile.alert.valid.description.preface"
```
Comment (deepsidhu85) [inline src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx:146]: Can we bold the preface and the postface? You can use the Typography component for this

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
A few things about the latest changes below

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx (hunk 7 above)
Comment (ksierks) [reply src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx]: Please see a54ec79fd261822b9a8c159c45f7c4db4a1fb00e.

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx (hunk 6 above)
Comment (ksierks) [reply src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx]: Please see a54ec79fd261822b9a8c159c45f7c4db4a1fb00e.

File: src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx (hunk 5 above)
Comment (ksierks) [reply src/main/webapp/resources/js/pages/projects/samples-metadata-import/components/SampleMetadataImportSelectFile.tsx]: Please see a54ec79fd261822b9a8c159c45f7c4db4a1fb00e.

**Review by deepsidhu85 — APPROVED** [review]
